File: examples/basic-window.c
/* basic-window.c
 *
 * Copyright 2025 Zach Podbielniak
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 *
 * Basic window example demonstrating Graylib usage.
 *
 * This example shows how to:
 * - Create a window
 * - Set up a basic game loop
 * - Clear the background with a color
 * - Draw simple text-like content (FPS counter)
 */

#include <graylib.h>
#include <stdio.h>

int
main (int   argc,
      char *argv[])
{
    g_autoptr(GrlWindow) window = NULL;
    g_autoptr(GrlColor) bg_color = NULL;
    g_autoptr(GrlColor) fg_color = NULL;
    g_autoptr(GrlVector2) center = NULL;
    gchar fps_text[32];
    gint frame_count;

    /* Create the window */
    window = grl_window_new (800, 600, "Graylib - Basic Window Example");

    if (!grl_window_is_ready (window))
    {
        g_printerr ("Failed to create window\n");
        return 1;
    }

    /* Set target FPS */
    grl_window_set_target_fps (window, 60);

    /* Create colors */
    bg_color = grl_color_new (40, 40, 60, 255);
    fg_color = grl_color_new_raywhite ();

    frame_count = 0;

    /* Main game loop */
    while (!grl_window_should_close (window))
    {
        gfloat delta_time;
        gint fps;
        gint width, height;

        /* Get timing info */
        delta_time = grl_window_get_frame_time (window);
        fps = grl_window_get_fps (window);
        frame_count++;

        /* Get window dimensions */
        width = grl_window_get_width (window);
        height = grl_window_get_height (window);

        /* Calculate center for drawing */
        if (center != NULL)
        {
            grl_vector2_free (center);
        }
        center = grl_vector2_new ((gfloat)width / 2.0f, (gfloat)height / 2.0f);

        /* Begin drawing */
        grl_window_begin_drawing (window);

        /* Clear background */
        grl_window_clear_background (window, bg_color);

        /* Draw a circle at the center */
        grl_draw_circle_v (center, 100.0f, fg_color);

        /* Draw some rectangles around the edges */
        {
            g_autoptr(GrlColor) red = grl_color_new_red ();
            g_autoptr(GrlColor) green = grl_color_new_green ();
            g_autoptr(GrlColor) blue = grl_color_new_blue ();
            g_autoptr(GrlColor) yellow = grl_color_new_yellow ();

            /* Corner rectangles */
            grl_draw_rectangle (10, 10, 50, 50, red);
            grl_draw_rectangle (width - 60, 10, 50, 50, green);
            grl_draw_rectangle (10, height - 60, 50, 50, blue);
            grl_draw_rectangle (width - 60, height - 60, 50, 50, yellow);
        }

        /* Draw a rotating polygon at center */
        {
            g_autoptr(GrlColor) poly_color = grl_color_new (200, 100, 255, 180);
            gfloat rotation = (gfloat)frame_count * 0.5f;

            grl_draw_poly (center, 6, 80.0f, rotation, poly_color);
        }

        /* End drawing */
        grl_window_end_drawing (window);

        /* Print FPS to terminal occasionally */
        if (frame_count % 60 == 0)
        {
            g_print ("FPS: %d, Delta: %.4f\n", fps, delta_time);
        }
    }

    g_print ("Window closed after %d frames\n", frame_count);

    return 0;
}

File: examples/basic-window.py
#!/usr/bin/env python3
# basic-window.py
#
# Copyright 2025 Zach Podbielniak
#
# SPDX-License-Identifier: AGPL-3.0-or-later
#
# Basic window example using Graylib Python bindings via GObject Introspection.
#
# Requirements:
#   - PyGObject (python3-gobject on Fedora, python3-gi on Debian/Ubuntu)
#   - Graylib library built with GIR support (make gir)
#
# Usage:
#   GI_TYPELIB_PATH=../build/gir LD_LIBRARY_PATH=../build/lib python3 basic-window.py
#
# Or after installation:
#   python3 basic-window.py

import gi
gi.require_version('Graylib', '1')
from gi.repository import Graylib as Grl

def main() -> int:
    """
    Basic window example demonstrating Graylib Python bindings.

    Creates a window and displays a simple animation with shapes.
    """
    # Initialize application
    app: Grl.Application = Grl.Application.new("com.example.basic-window-py")

    # Create window
    window: Grl.Window = Grl.Window.new(800, 600, "Graylib - Python Example")
    window.set_target_fps(60)

    # Create colors
    bg_color: Grl.Color = Grl.Color.new(40, 40, 60, 255)
    white: Grl.Color = Grl.Color.new(255, 255, 255, 255)
    red: Grl.Color = Grl.Color.new(230, 41, 55, 255)
    green: Grl.Color = Grl.Color.new(0, 228, 48, 255)
    blue: Grl.Color = Grl.Color.new(0, 121, 241, 255)

    # Animation state
    ball_x: float = 400.0
    ball_y: float = 300.0
    ball_vx: float = 200.0
    ball_vy: float = 150.0
    ball_radius: float = 30.0

    # Main loop
    while not window.should_close():
        # Get frame time
        dt: float = window.get_frame_time()

        # Update ball position
        ball_x += ball_vx * dt
        ball_y += ball_vy * dt

        # Bounce off walls
        if ball_x - ball_radius < 0 or ball_x + ball_radius > 800:
            ball_vx = -ball_vx
            ball_x = max(ball_radius, min(800 - ball_radius, ball_x))

        if ball_y - ball_radius < 0 or ball_y + ball_radius > 600:
            ball_vy = -ball_vy
            ball_y = max(ball_radius, min(600 - ball_radius, ball_y))

        # Begin drawing
        window.begin_drawing()
        Grl.Draw.clear_background(bg_color)

        # Draw title
        Grl.Draw.text("Graylib - Python GI Bindings Example", 10, 10, 20, white)
        Grl.Draw.text("Bouncing ball animation", 10, 35, 16, white)

        # Draw bouncing ball
        Grl.Draw.circle(int(ball_x), int(ball_y), ball_radius, red)
        Grl.Draw.circle_lines(int(ball_x), int(ball_y), ball_radius, white)

        # Draw some static shapes
        Grl.Draw.rectangle(50, 500, 100, 50, green)
        Grl.Draw.rectangle_lines(50, 500, 100, 50, white)

        Grl.Draw.rectangle(650, 500, 100, 50, blue)
        Grl.Draw.rectangle_lines(650, 500, 100, 50, white)

        # Draw FPS
        Grl.Draw.fps(10, 570)

        window.end_drawing()

    return 0


if __name__ == "__main__":
    import sys
    sys.exit(main())
